examples/bridge-app/linux/include/Clusters.h

```diff
@@ -0,0 +1,294 @@
+        if (kIsChar)
+            return aEncoder.Encode(chip::CharSpan(static_cast<char *>(mValue), mLength));
+        return aEncoder.Encode(chip::ByteSpan(mValue, mLength));
+    }
+
+    chip::ByteSpan Peek() const { return IsNull() ? chip::ByteSpan() : chip::ByteSpan(mValue, mLength); }
+
+    bool IsNull() const { return mLength == 0xFFFF; }
+
+private:
+    uint16_t mLength = 0xFFFF;
+    uint8_t mValue[kMaxSize];
+};
+
+// This allows for an array of up to MaxElements of a specified type.
+template <size_t MaxElements, typename T>
+struct ArrayType
+{
+    static constexpr EmberAfAttributeType kMatterType = ZCL_ARRAY_ATTRIBUTE_TYPE;
+    static constexpr uint16_t kMaxSize                = sizeof(T) * MaxElements;
+    static_assert(std::is_standard_layout<T>::value, "Array element not standard layout!");
+
+    void ListWriteBegin(const chip::app::ConcreteAttributePath & aPath) {}
+
+    void ListWriteEnd(const chip::app::ConcreteAttributePath & aPath, bool aWriteWasSuccessful) {}
+
+    CHIP_ERROR Write(const chip::app::ConcreteDataAttributePath & aPath, chip::app::AttributeValueDecoder & aDecoder)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Read(const chip::app::ConcreteReadAttributePath & aPath, chip::app::AttributeValueEncoder & aEncoder)
+    {
+        if (aPath.mListIndex.HasValue())
+        {
+            uint16_t index = aPath.mListIndex.Value();
+            if (index >= mLength || IsNull())
+                return CHIP_ERROR_INVALID_ARGUMENT;
+
+            return aEncoder.Encode(mArray[index]);
+        }
+
+        if (IsNull())
+            return aEncoder.EncodeNull();
+        if (mLength == 0)
+            return aEncoder.EncodeEmptyList();
+        return aEncoder.EncodeList([this](const auto & encoder) {
+            CHIP_ERROR err = CHIP_NO_ERROR;
+            for (uint16_t i = 0; i < mLength && err == CHIP_NO_ERROR; i++)
+                err = encoder.Encode(mArray[i]);
+            return err;
+        });
+    }
+
+    bool IsNull() const { return mLength == 0xFFFF; }
+
+private:
+    uint16_t mLength = 0xFFFF;
+    T mArray[MaxElements];
+};
+
+// This is used by code generation to provide specific attributes. Type should be one of
+// PrimitiveType, OctetString, ArrayType, etc to provide storage and type-specific
+// functionality.
+template <chip::AttributeId id, EmberAfAttributeMask mask, typename Type>
+struct Attribute : public Type
+{
+    static constexpr chip::AttributeId kId    = id;
+    static constexpr EmberAfClusterMask kMask = mask | ZAP_ATTRIBUTE_MASK(EXTERNAL_STORAGE);
+
+    using Type::Type;
+    using Type::operator=;
+};
+
+struct CommonCluster;
+typedef std::function<CHIP_ERROR(CommonCluster *, const chip::app::ConcreteDataAttributePath &, chip::app::AttributeValueDecoder &)>
+    PropagateWriteCB;
+
+// This is the base type of all generated clusters, providing the backend access to
+// implementation details.
+struct CommonCluster : public ClusterImpl
```

Comment:
The namings here are a bit odd - `ClusterImpl` is an interface, but it contains the term 'Impl' in it.

Recommendation: Rename to `CommonClusterImpl` and `ClusterInterface`.